deploy/buildtools/buildfarm.py

```diff
@@ -0,0 +1,334 @@
+import logging
+import sys
+import abc
+
+from awstools.awstools import *
+
+# imports needed for python type checking
+from typing import cast, Any, Dict, Optional, Sequence, List, TYPE_CHECKING
+from mypy_boto3_ec2.service_resource import Instance as EC2InstanceResource
+# needed to avoid type-hint circular dependencies
+# TODO: Solved in 3.7.+ by "from __future__ import annotations" (see https://stackoverflow.com/questions/33837918/type-hints-solve-circular-dependency)
+#       and normal "import <module> as ..." syntax (see https://www.reddit.com/r/Python/comments/cug90e/how_to_not_create_circular_dependencies_when/)
+if TYPE_CHECKING:
+    from buildtools.buildconfig import BuildConfig
+else:
+    BuildConfig = object
+
+rootLogger = logging.getLogger()
+
+class BuildHost:
+    """Abstract class representing a single build host which holds a single build config.
+
+    Attributes:
+        build_config: Build config associated with the build host.
+        dest_build_dir: Name of build dir on build host.
+        ip_address: IP address of build host.
+    """
+    build_config: BuildConfig
+    dest_build_dir: str
+    ip_address: str
+
+    def __init__(self):
+        self.ip_address = ""
+        self.build_config = None # type: ignore
+        self.dest_build_dir = ""
+
+class BuildFarm(metaclass=abc.ABCMeta):
+    """Abstract class representing a build farm managing multiple build hosts (request, wait, release, etc).
+
+    Attributes:
+        build_hosts: List of build hosts used for builds.
+        args: Set of args/options associated with the build farm.
+    """
+    build_hosts: List[BuildHost]
+    args: Dict[str, Any]
+
+    def __init__(self, args: Dict[str, Any]) -> None:
+        """
+        Args:
+            args: Args (i.e. options) passed to the build farm.
+        """
+        self.args = args
+        self.build_hosts = []
+
+    @staticmethod
+    @abc.abstractmethod
+    def NAME() -> str:
+        """Human-readable name (used as the "build-farm-type" in the YAML).
+
+        Returns:
+            Human-readable name.
+        """
+        raise NotImplementedError
```

Comment:
I don't know. I'll check

Reply:
Technically you don't need to do the `raise ...`. You can just `return` (since you need something in the function body) and the decorator will cause an error like `TypeError: Can't instantiate abstract class CHILDCLASS with abstract methods ABSTRACTMETHOD`. I'll change this around.

Reply:
Not a dumb question.  ABC in python is weird.  It requires that subclasses implement the method but the base class can also have behavior that can only be explicitly invoked via `super()`.  From the end of `@abstractmethod` docs at https://docs.python.org/3/library/abc.html#abc.abstractmethod
> Note Unlike Java abstract methods, these abstract methods may have an implementation. This implementation can be called via the [super()](https://docs.python.org/3/library/functions.html#super) mechanism from the class that overrides it. This could be useful as an end-point for a super-call in a framework that uses cooperative multiple-inheritance.

Reply:
So, I had said to Abe that it is probably good to explicitly throw so that people don't think there's some kind of behavior happening and to make it explicit that we're not using python's interesting feature.

Reply:
@abejgonzalez , I see that you took the raise statements out.  It's probably fine.  Don't go back and undo the removal.